sdks/python/apache_beam/io/datastore/v1/datastoreio.py

```diff
@@ -0,0 +1,289 @@
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""A source and a sink for reading from and writing to Google Cloud Datastore"""
+
+import logging
+
+from googledatastore import helper as datastore_helper
+
+from apache_beam.io.datastore.v1 import helper
+from apache_beam.io.datastore.v1 import query_splitter
+from apache_beam.transforms import Create
+from apache_beam.transforms import DoFn
+from apache_beam.transforms import FlatMap
+from apache_beam.transforms import GroupByKey
+from apache_beam.transforms import PTransform
+from apache_beam.transforms import ParDo
+from apache_beam.transforms.util import Values
+
+__all__ = ['ReadFromDatastore']
+
+
+class ReadFromDatastore(PTransform):
+  """A ``PTransform`` for reading from Google Cloud Datastore.
+
+  To read a `PCollection[Entity]` from a Cloud Datastore `Query`, use
+  `ReadFromDatastore` transform by providing a `project` id and a `query` to
+  read from. You can optionally provide a `namespace` and/or specify how many
+  splits you want for the query through `num_splits` option.
+
+  Note: Normally, a Runner will read from Cloud Datastore in parallel across
+  many workers. However, when the `query` is configured with a `limit` or if the
+  query contains inequality filters like `GREATER_THAN, LESS_THAN` etc., then
+  all the returned results will be read by a single worker in order to ensure
+  correct data. Since data is read from a single worker, this could have
+  significant impact on the performance of the job.
+
+  The semantics for the query splitting is defined below:
+    1. Any value for `num_splits` less than or equal to 0 will be ignored, and
+    the number of splits will be chosen dynamically at runtime based on the
+    query data size.
+
+    2. Any value ofr `num_splits` greater than
+    `ReadFromDatastore._NUM_QUERY_SPLITS_MAX` will be capped at that value.
+
+    3. If the `query` has a user limit set, or contains inequality filters, then
+    `num_splits` will be ignored and no split will be performed.
+
+    4. Under certain cases Cloud Datastore is unable to split query to the
+    requested number of splits. In such cases we just use whatever the Cloud
+    Datastore returns.
+
+  See https://developers.google.com/datastore/ for more details on Google Cloud
+  Datastore.
+  """
+
+  # An upper bound on the number of splits for a query.
+  _NUM_QUERY_SPLITS_MAX = 50000
```

Comment:
These are just rough estimates and there is no strong reason for choosing that exact number. 12 because is a good multiple of many numbers (so easy to uniformly parallelize with 2, 3, 4, 6 workers). Practically, 50000 splits should be good enough to parallelize any query, it is just an upper bound but Cloud Datastore really controls how many splits it can give. Keeping it consistent with java, we could revisit it later if needed.